velox/expression/tests/ArgGenerator.h

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include "velox/expression/FunctionSignature.h"
+#include "velox/vector/fuzzer/Utils.h"
+
+namespace facebook::velox::test {
+
+class ArgGenerator {
+ public:
+  virtual ~ArgGenerator() = default;
+
+  /// Given a signature and a concrete return type return randomly selected
+  /// valid input types. Returns empty vector if no input types can produce the
+  /// specified result type.
+  virtual std::vector<TypePtr> generateArgs(
+      const exec::FunctionSignature& signature,
+      const TypePtr& returnType,
+      FuzzerGenerator& rng) = 0;
+};
+
+/// An argument type generator for decimal function. A map keyed on the pair of
+/// precision and scale could be initialized with all possible input types.
+/// Argument types are generated by looking up the map with the precision and
+/// scale of return type, and randomly selecting valid input types.
+class DecimalArgGeneratorBase : public ArgGenerator {
+ public:
+  // Generate decimal argument types from a given signature and return type.
+  std::vector<TypePtr> generateArgs(
+      const exec::FunctionSignature& /*signature*/,
+      const TypePtr& returnType,
+      FuzzerGenerator& rng) override;
+
+ protected:
+  using Inputs = std::vector<TypePtr>;
+
+  /// Return randomly selected pair of input types that produce the specified
+  /// result type.
+  Inputs findInputs(const TypePtr& returnType, FuzzerGenerator& rng) const;
+
+  // Return all the possible decimal types.
+  const std::vector<TypePtr>& getAllTypes() const;
+
+  /// Compute result type for all possible pairs of decimal input types. Store
+  /// the results in 'inputs_' maps keyed by the precision and scale of return
+  /// type. By default, the result type is considered to be calculated from two
+  /// argument decimal types.
+  virtual void initialize();
+
+  /// Given precisions and scales of the inputs, return precision and scale of
```

Comment:
Do we need to define what return value is if inputs are not "valid / supported"?

Reply:
Added below documentation.
> Returns std::nullopt if a valid return type cannot be generated with inputs.